mirrord-config/derive/src/lib.rs

```diff
@@ -0,0 +1,239 @@
+    let Field { attrs, ident, .. } = field;
+
+    let flags = attrs
+        .iter()
+        .find(|attr| attr.path.is_ident("config"))
+        .and_then(|attr| attr.parse_meta().ok())
+        .map(get_config_flags)
+        .unwrap_or_default();
+
+    let mut impls = Vec::new();
+
+    let env_flag = flags.iter().fold(None, |acc, flag| match flag {
+        FieldFlags::Env(env) => Some(env),
+        _ => acc,
+    });
+
+    if let Some(flag) = env_flag {
+        impls.push(quote! { crate::config::from_env::FromEnv::new(#flag) });
+    }
+
+    if flags.contains(&FieldFlags::Nested) {
+        impls.push(quote! { Some(self.#ident.generate_config()?) })
+    } else {
+        impls.push(quote! { self.#ident.clone() });
+    }
+
+    if let Some(FieldFlags::Default(flag)) = flags
+        .iter()
+        .find(|flag| matches!(flag, FieldFlags::Default(_)))
+    {
+        impls.push(quote! { crate::config::default_value::DefaultValue::new(#flag) });
+    }
+
+    let unwrapper = flags.iter().find(|flag| matches!(flag, FieldFlags::Unwrap | FieldFlags::Nested | FieldFlags::Default(_))).map(
+        |_| {
+            let env_override = match env_flag {
+                Some(flag) => quote! { Some(#flag) },
+                None => quote! { None }
+            };
+            quote! { .ok_or(crate::config::ConfigError::ValueNotProvided(stringify!(#parent), stringify!(#ident), #env_override))? }
+        }
+    );
+
+    quote! {
+        #ident: (#(#impls),*).source_value() #unwrapper
+    }
+}
+
+#[proc_macro_derive(MirrordConfig, attributes(config))]
+pub fn mirrord_config(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let DeriveInput {
+        attrs,
+        ident,
+        data,
+        generics,
+        vis,
+    } = syn::parse_macro_input!(input as DeriveInput);
+
+    let mapped_name = map_to_ident(
+        &ident,
+        attrs
+            .iter()
+            .find(|attr| attr.path.is_ident("config"))
+            .and_then(|attr| attr.parse_args::<Expr>().ok()),
+    );
+
+    let mapped_type = match data {
+        Data::Struct(_) => quote! { struct },
+        Data::Enum(_) => quote! { enum },
+        _ => panic!("Unions are not supported"),
+    };
+
+    let fields = match data {
+        Data::Struct(DataStruct { fields, .. }) => match fields {
+            Fields::Named(FieldsNamed { named, .. }) => named,
+            _ => todo!(),
+        },
+        _ => todo!(),
+    };
+
+    let mapped_fields = {
+        let named = fields
+            .clone()
+            .into_iter()
+            .map(map_field_name)
+            .collect::<Vec<_>>();
+
+        quote! { { #(#named),* } }
+    };
+
+    let mapped_fields_impl = {
+        let named = fields
+            .into_iter()
+            .map(|field| map_field_name_impl(&ident, field))
+            .collect::<Vec<_>>();
+
+        quote! { { #(#named),* } }
+    };
+
+    let output = quote! {
+        #[derive(Clone, Debug)]
```

Comment:
Shouldn't we use the original derives? Like if the user had:

```rust
#[derive(Clone, PartialEq, Foo, MirrordConfig)]
struct Foo ...
```

Doesn't this macro kinda eat up the other derives?

Reply:
Same applies for the doc comments added to the original struct, aren't they lost here?

Reply:
About the macros not sure how to get the other macros, I printed the macro input and didn't get other derives, also as far as I see derive macros don't touch the original struct so no other derives are overwritten just don't get copied over to Mapped struct

Reply:
Oh, thought that `derive` behaved like the other macros, mb.